crates/uv/src/commands/auth/helper.rs

```diff
@@ -0,0 +1,143 @@
+            .to_owned();
+
+        Ok(Self {
+            headers: HashMap::from([("Authorization".to_owned(), vec![header_str])]),
+        })
+    }
+
+    type Error = anyhow::Error;
+}
+
+async fn credentials_for_url(
+    url: &DisplaySafeUrl,
+    preview: Preview,
+    network_settings: &NetworkSettings,
+) -> Result<Option<Credentials>> {
+    let pyx_store = PyxTokenStore::from_settings()?;
+
+    // Use only the username from the URL, if present - discarding the password
+    let url_credentials = Credentials::from_url(url);
+    let username = url_credentials.as_ref().and_then(|c| c.username());
+    if url_credentials
+        .as_ref()
+        .map(|c| c.password().is_some())
+        .unwrap_or(false)
+    {
+        debug!("URL '{url}' contain a password; ignoring");
+    }
+
+    if pyx_store.is_known_domain(url) {
+        if username.is_some() {
+            bail!(
+                "Cannot specify a username for URLs under {}",
+                url.host()
+                    .map(|host| host.to_string())
+                    .unwrap_or(url.to_string())
+            );
+        }
+        let client = BaseClientBuilder::new(
+            network_settings.connectivity,
+            network_settings.native_tls,
+            network_settings.allow_insecure_host.clone(),
+            preview,
+            network_settings.timeout,
+            network_settings.retries,
+        )
+        .auth_integration(uv_client::AuthIntegration::NoAuthMiddleware)
+        .build();
+        let maybe_token = pyx_store
+            .access_token(client.for_host(pyx_store.api()).raw_client(), 0)
+            .await
+            .context("Authentication failure")?;
+        let token = maybe_token.context("No access token found")?;
+        return Ok(Some(Credentials::bearer(token.into_bytes())));
+    }
+    let backend = AuthBackend::from_settings(preview).await?;
+    let credentials = match &backend {
+        AuthBackend::System(provider) => provider.fetch(url, username).await,
+        AuthBackend::TextStore(store, _lock) => store.get_credentials(url, username).cloned(),
+    };
+    Ok(credentials)
+}
+
+/// Implement the Bazel credential helper protocol.
+///
+/// Reads a JSON request from stdin containing a URI, looks up credentials
+/// for that URI using uv's authentication backends, and writes a JSON response
+/// to stdout containing HTTP headers (if credentials are found).
+///
+/// Protocol specification TLDR:
+/// - Input (stdin): `{"uri": "https://example.com/path"}`
+/// - Output (stdout): `{"headers": {"Authorization": ["Basic ..."]}}` or `{"headers": {}}`
+/// - Errors: Written to stderr with non-zero exit code
+///
+/// Full spec is [available here](https://github.com/bazelbuild/proposals/blob/main/designs/2022-06-07-bazel-credential-helpers.md)
+pub(crate) async fn helper(
+    preview: Preview,
+    network_settings: &NetworkSettings,
+    printer: Printer,
+) -> Result<ExitStatus> {
```

Comment:
I think I'd emit a preview warning for this command / have a feature flag for it. In general, I'm wary to add new interfaces without going through preview to start so we can make breaking changes if we need to.